### PR TITLE
Thunderscope: Dynamic Parameters Reborn + Port to Tscope

### LINKED ISSUE #1550: DynamicParameters: Register callback when any parameter in a config changes
### Description of the task

From slack

> Being able to register a callback for when any value in an entire config changes. Eg I want to re-init networking code when a networking param changes, but I don't want to enumerate all the parameters and register a callback for each

No strong preference for if this is for the parameters that are immediate children of the config, of if this recursively applies to all subconfigs too. Whichever is easier, I can't think of a use-case that demands one or the other atm.

This would be a nice quality-of-life improvement and help keep some code smaller/cleaner by not manually registering a function to a bunch of params.

### Acceptance criteria

- [ ] Allow callbacks to be called when any parameter in a config changes

### LINKED ISSUE #1738: DynamicParameters: Document Parameter/Config Usage and Conventions
### Description of the task

Currently no documents exist on how to use dynamic parameters and what conventions we follow when we create new parameters/configs. This ticket involves writing documentation to cover that, once the new parameter system is implemented.

### Acceptance criteria

- [ ] Write dynamic parameter docs
- [ ] Write usage/conventions docs
- [ ] Cover command line args support

### Blocked By
#1298

### LINKED ISSUE #2564: Thunderscope: Create Dynamic Parmeter Widget From Proto

### PR DESCRIPTION
**Pull request by akhilveeraghanta**
### Description

| Removed | Replaced With|
| --- | --- |
| Dynamic Parameter Yamls | parameters.proto|
| Full System GUI  | Thunderscope |
| Standalone Sim GUI | Thunderscope |
| Robot Diagnostics | Thunderscope TODO #2563 | 
| Hand Held Controlelr | Thunderscope TODO #2562 | 
| all of firmware | Thunderloop |

| New Feature | Description | 
|  --- | --- |
| ProtoConfigurationWidget | Creates a widget from a protobuf, useful to create a searchable widget from _any_ protobuf |
| parameters.proto | Defines all the configs + parameters and their ranges + acceptable values |

Instead of looking at yaml files to generate parameters that create a giant tree with share ptrs, 
we looked at a protobuf that defines config msgs and dependency inject that through our code base

We can iterate over the config and generate a searchable parameter widget

### Testing Done
Manual testing, existing tests pass

### Resolved Issues
resolves #2564 
resolves #1550
resolves #1738

### Review Checklist
- [ ] src/software/sensor_fusion/sensor_fusion.cpp
- [ ] src/software/simulated_tests/simulated_er_force_sim_play_test_fixture.cpp
- [ ] src/software/simulated_tests/simulated_er_force_sim_test_fixture.cpp
- [ ] src/software/simulated_tests/simulated_test_fixture.py
- [ ] src/software/thunderscope/common/proto_configuration_widget.py
- [ ] src/software/ai/threaded_ai.cpp
- [ ] src/software/ai/ai.cpp
- [ ] src/software/unix_full_system_main.cpp

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

File: formatting_scripts/fix_formatting.sh [hunk 1]
```diff
@@ -68,8 +68,8 @@ function run_python_import_formatting(){
 # Function to run black python formatting
 function run_black_formatting () {
     printf "Running black to format Python files...\n\n"
-    # suppress misleading "All done! ✨ 🍰 ✨" message
-    $CURR_DIR/black_$BLACK_VERSION $BAZEL_ROOT_DIR &>/dev/null
+
+    $CURR_DIR/black_$BLACK_VERSION $BAZEL_ROOT_DIR
```
Comment (jonathanlew) [inline formatting_scripts/fix_formatting.sh:72]: why are there changes to this file?

File: src/software/ai/ai.h [hunk 2]
```diff
@@ -48,24 +48,31 @@ class AI final
...
+    /**
+     * Update the AiConfig proto
+     *
+     * @param ai_config The new AiConfig proto
+     */
+    void updateAiConfig(TbotsProto::AiConfig ai_config);
+
+   private:
     void checkAiConfig();
+    bool ai_config_changed = false;
```
Comment (jonathanlew) [inline src/software/ai/ai.h]: this should be initialized in the constructor

File: src/software/ai/ai.cpp [hunk 3]
```diff
@@ -20,39 +20,55 @@ void AI::overridePlay(std::unique_ptr<Play> play)
...
 
-    if (!field_to_path_planner_factory.contains(world.field()))
+    if (!field_to_path_planner_factory.contains(world.field()) || ai_config_changed)
     {
-        field_to_path_planner_factory.emplace(
+        field_to_path_planner_factory.insert_or_assign(
             world.field(),
-            GlobalPathPlannerFactory(ai_config->getRobotNavigationObstacleConfig(),
+            GlobalPathPlannerFactory(ai_config_.robot_navigation_obstacle_config(),
                                      world.field()));
```
Comment (jonathanlew) [inline src/software/ai/ai.cpp:80]: a comment would help here

File: src/software/ai/ai.cpp [hunk 4]
```diff
@@ -20,39 +20,55 @@ void AI::overridePlay(std::unique_ptr<Play> play)
...
+    }
+
+    // If we have a new override but its back to the Ai selection, lets
+    // clear the override
+    if (current_override != prev_override &&
+        current_override == TbotsProto::PlayName::UseAiSelection)
     {
-        overridePlayFromName(
-            ai_config->getAiControlConfig()->getCurrentAiPlay()->value());
+        overridePlay(nullptr);
```
Comment (jonathanlew) [inline src/software/ai/ai.cpp]: why do we need both `overridePlayFromProto` and `overridePlay`?

File: src/software/ai/hl/stp/tactic/crease_defender/crease_defender_fsm.h [hunk 5]
```diff
@@ -78,6 +78,8 @@ struct CreaseDefenderFSM
     }
 
    private:
+    double robot_obstacle_inflation_factor_;
```
Comment (jonathanlew) [inline src/software/ai/hl/stp/tactic/crease_defender/crease_defender_fsm.h]: I don't see this being used

**Review by jonathanlew — COMMENTED** [review]
Brilliant! Will have to check out the dynamic param widget

File: formatting_scripts/fix_formatting.sh (hunk 1 above)
Comment (akhilveeraghanta) [reply formatting_scripts/fix_formatting.sh:72]: we suppressed the black logs, but it actually has really good outputs which tell you exactly where the syntax parsing is failing, which is useful in finding where you are missing a bracket in python code

File: src/software/ai/ai.cpp (hunk 4 above)
Comment (akhilveeraghanta) [reply src/software/ai/ai.cpp]: There are some instances in simulated play test fixtures that use overridePlay to create a play, update control params and then override. We also call it from the overridePlayFromProto function (which relaces the previous overridePlayFromName)

We use overridePlayFromProto for the simulated playtests

File: src/software/ai/hl/stp/tactic/crease_defender/crease_defender_fsm.h (hunk 5 above)
Comment (akhilveeraghanta) [reply src/software/ai/hl/stp/tactic/crease_defender/crease_defender_fsm.h]: removed

File: src/software/ai/ai.cpp (hunk 3 above)
Comment (akhilveeraghanta) [reply src/software/ai/ai.cpp:80]: done

File: src/software/ai/ai.h (hunk 2 above)
Comment (akhilveeraghanta) [reply src/software/ai/ai.h]: done

**Review by jonathanlew — COMMENTED** [review]
reminder to make sure the searched for dynamic parameter widgets still work

**Comment by akhilveeraghanta** [comment]
> reminder to make sure the searched for dynamic parameter widgets still work

completely forgot thanks for reminder, fixed!

File: src/software/ai/ai.h [hunk 6]
```diff
@@ -48,24 +48,31 @@ class AI final
...
-    std::shared_ptr<const AiConfig> ai_config;
+    TbotsProto::AiConfig ai_config_;
     std::unique_ptr<FSM<PlaySelectionFSM>> fsm;
     std::unique_ptr<Play> override_play;
     std::unique_ptr<Play> current_play;
     std::map<Field, GlobalPathPlannerFactory> field_to_path_planner_factory;
-    bool prev_override;
-    std::string prev_override_name;
+    TbotsProto::PlayName prev_override;
+    bool ai_config_changed;
```
Comment (jonathanlew) [inline src/software/ai/ai.h:74]: doesn't look like it's initialized in the constructor

Comment (akhilveeraghanta) [reply src/software/ai/ai.h:74]: hmm i see it in ais constructor (both prev_override and ai_config_changed)

**Review by jonathanlew — DISMISSED** [review]
LGTM! Nice work! :microscope:

File: src/software/thunderscope/common/proto_configuration_widget.py [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+            limits=options,
+        )
+
+    @staticmethod
+    def __create_bool_parameter(key, value, descriptor):
+        """Convert a bool field in proto to a BoolParameter
+
+        :param key: The name of the parameter
+        :param value: The default value
+        :param _: The proto descriptor, unused for bool
```
Comment (nimazareian) [inline src/software/thunderscope/common/proto_configuration_widget.py]: nit: I think here the `descriptor` argument should be `_`, or the comment `_` be `descriptor` so its slightly more clear

File: src/software/thunderscope/common/proto_configuration_widget.py [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+    @staticmethod
+    def __create_string_parameter(key, value, descriptor):
+        """Convert a bool field in proto to a BoolParameter
+
+        :param key: The name of the parameter
+        :param value: The default value
+        :param descriptor: The proto descriptor
+
+        """
+        return {"name": key, "type": "bool", "value": value}
```
Comment (nimazareian) [inline src/software/thunderscope/common/proto_configuration_widget.py]: This seems pretty much identical to `__create_bool_parameter`. I'm also curious why the `"type"` is "bool" for strings? Is this meant to be like this, or is it an accidental duplication?

File: src/software/thunderscope/common/proto_configuration_widget.py [hunk 9]
```diff
@@ -0,0 +1,306 @@
...
+        that can loaded directly into a ParameterTree
+
+        https://pyqtgraph.readthedocs.io/en/latest/parametertree/index.html
+
+        :param message: The message to convert to a dictionary
+        :param search_term: The search filter
+        :param current_attr: Which attr we are currently on to set
+        :param convert_all_fields_to_bools: 
+                        Sometimes we just want to check/uncheck the fields
+                        of the protobuf (and not actually set an explicit value)
```
Comment (nimazareian) [inline src/software/thunderscope/common/proto_configuration_widget.py]: I'm a bit confused on when we would use `convert_all_fields_to_bools`. 
e.g. what does it mean for `robot_obstacle_inflation_factor` when `convert_all_fields_to_bools` is true?

**Review by nimazareian — COMMENTED** [review]
Just tried it and it's super cool. It is also much easier to understand than the previous dynamic parameter implementation!
Though I'm a bit confused on where we do the communication between python and C++ when values are updated. Edit: Actually seems to be [here](https://github.com/UBC-Thunderbots/Software/blob/5ac9da7f8673bc84a0ae0adb8bd5ff225fdc137b/src/software/thunderscope/thunderscope.py#L459).

File: src/software/ai/ai.cpp [hunk 10]
```diff
@@ -20,39 +21,58 @@ void AI::overridePlay(std::unique_ptr<Play> play)
...
+    // We construct the global path planner once for the field. If the AI Config
+    // changes, there might be an update we need to reconstruct the path planner
+    // for and propagate the parameter change.
+    if (!field_to_path_planner_factory.contains(world.field()) || ai_config_changed)
     {
-        field_to_path_planner_factory.emplace(
+        field_to_path_planner_factory.insert_or_assign(
             world.field(),
-            GlobalPathPlannerFactory(ai_config->getRobotNavigationObstacleConfig(),
+            GlobalPathPlannerFactory(ai_config_.robot_navigation_obstacle_config(),
```
Comment (nimazareian) [inline src/software/ai/ai.cpp:79]: Is the `GlobalPathPlannerFactory` the only class that needs to be updated with the new config? Do we not need to update any of the plays (i.e. the current play)? Or is the idea that once we switch to a new play and it's constructed, then it will get the updated config values?

File: src/software/thunderscope/common/proto_configuration_widget.py [hunk 11]
```diff
@@ -0,0 +1,306 @@
...
+        # Create search query bar
+        self.search_query = QLineEdit()
+        self.search_query.textChanged.connect(self.__handle_search_query_changed)
+        self.search_filter_threshold = search_filter_threshold
+
+        layout.addWidget(self.search_query)
+        layout.addWidget(self.param_tree)
+
+    def __handle_search_query_changed(self, search_term):
+        """Given a new search term, reconfigure the parameter tree with parameters
```
Comment (itsarune) [inline src/software/thunderscope/common/proto_configuration_widget.py:62]: nit: with parameters that match the term

File: src/software/ai/hl/stp/stp_tactic_assignment_test.cpp [hunk 12]
```diff
@@ -18,18 +18,13 @@
 class STPTacticAssignmentTest : public ::testing::Test
 {
    public:
-    STPTacticAssignmentTest()
-        : thunderbots_config(std::make_shared<const ThunderbotsConfig>()),
-          stp(thunderbots_config->getAiConfig())
-    {
-    }
+    STPTacticAssignmentTest() : ai_config(TbotsProto::AiConfig()), stp(ai_config) {}
```
Comment (itsarune) [inline src/software/ai/hl/stp/stp_tactic_assignment_test.cpp:21]: why does this file still exist, doesn't Jon's pr remove this?

**Review by itsarune — COMMENTED** [review]
lgtm! looks like it's working on my computer and responds to changes via param searches

File: src/software/ai/hl/stp/stp_tactic_assignment_test.cpp (hunk 12 above)
Comment (akhilveeraghanta) [reply src/software/ai/hl/stp/stp_tactic_assignment_test.cpp:21]: good catch, seems like they were missed in that PR

I don't know which tests to keep and which to remove so ill leave it to @jonathanlew to open another PR moving/removing these files

Comment (akhilveeraghanta) [reply src/software/ai/hl/stp/stp_tactic_assignment_test.cpp:21]: they aren't in the BUILD file so they don't run 
and maybe we want to keep the play selection fsm test

File: src/software/thunderscope/common/proto_configuration_widget.py (hunk 7 above)
Comment (akhilveeraghanta) [reply src/software/thunderscope/common/proto_configuration_widget.py]: fixed

File: src/software/thunderscope/common/proto_configuration_widget.py (hunk 8 above)
Comment (akhilveeraghanta) [reply src/software/thunderscope/common/proto_configuration_widget.py]: good catch! fixed

File: src/software/thunderscope/common/proto_configuration_widget.py (hunk 9 above)
Comment (akhilveeraghanta) [reply src/software/thunderscope/common/proto_configuration_widget.py]: yeah its bit of a weird feature to toss in here, its better for the other class to override the create_* functions and setup this bool feature so I've removed it from here

but its for the arbitrary plotter. we'd want the option to select floats as bools to plot (as opposed to setting the float itself)

File: src/software/ai/ai.cpp (hunk 10 above)
Comment (akhilveeraghanta) [reply src/software/ai/ai.cpp:79]: so the main modules we have that need to be updated are:

sensor_fusion
play_selection_fsm (which will handle tactics, plays, pass gen, etc..)
path_planner

sensor fusion will reset itself if the sensor fusion config changes
path planner reconstructs itself when the ai_config changes

and finally the playselection fsm will receive an update event with the ai_config

that being said, when I went to go take a closer look, the way I had it setup rn the play selection fsm would have only updated the ai_config when the play changed

**TL;DR I've added a line of code to restart the playselection fsm if the ai config change**s

File: src/software/thunderscope/common/proto_configuration_widget.py (hunk 11 above)
Comment (akhilveeraghanta) [reply src/software/thunderscope/common/proto_configuration_widget.py:62]: fixed

**Review by jonathanlew — DISMISSED** [review]

File: src/software/thunderscope/common/proto_configuration_widget.py (hunk 9 above)
Comment (nimazareian) [reply src/software/thunderscope/common/proto_configuration_widget.py]: Im still slightly confused on what you mean by select floats as bools to plot. Does that mean that floats would be converted to 0 or 1 when they're plotted?

File: src/software/thunderscope/common/proto_configuration_widget.py [hunk 13]
```diff
@@ -0,0 +1,285 @@
...
+        :param key: The name of the parameter
+        :param value: The default value
+        :param _: The proto descriptor, unused for bool
+
+        """
+        return {"name": key, "type": "bool", "value": value}
+
+    @staticmethod
+    def __create_string_parameter(key, value, descriptor):
+        """Convert a bool field in proto to a BoolParameter
```
Comment (nimazareian) [inline src/software/thunderscope/common/proto_configuration_widget.py]: nit: should be something like "Convert a string field in proto to a StrParameter"

**Review by nimazareian — CHANGES_REQUESTED** [review]
Looks good. Just one last nit

File: src/software/thunderscope/common/proto_configuration_widget.py (hunk 13 above)
Comment (akhilveeraghanta) [reply src/software/thunderscope/common/proto_configuration_widget.py]: good catch, fixed!

**Review by nimazareian — APPROVED** [review]
Looks great! :100:

**Review by jonathanlew — APPROVED** [review]
